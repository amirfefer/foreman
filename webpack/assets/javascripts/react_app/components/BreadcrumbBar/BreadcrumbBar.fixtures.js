export const breadcrumbItems = {
  items: [
    {
      caption: 'root',
      url: '/some-url',
    },
    {
      caption: 'child',
    },
  ],
};

export const breadcrumbTitleItems = {
  items: [
    {
      caption: 'title',
    },
  ],
};

export const resource = {
  resourceUrl: 'some/url',
  nameField: 'name',
  switcherItemUrl: 'some/url/:id',
};

export const resourceList = [
  { name: 'Host 1', url: '#' },
  { name: 'Host 2', url: '#' },
  { name: 'Host 3 with a very long name', url: '#' },
  { name: 'Host 4', url: undefined, onClick: jest.fn() },
  { name: 'Host 5', url: '#', onClick: undefined },
];

export const serverResourceListResponse = {
  data: {
    page: 1,
    total: 3,
    per_page: 2,
    results: [
      { id: '1', name: 'name-1' },
      { id: '2', name: 'name-2' },
      { id: '3', name: 'name-3' },
    ],
  },
};

export const breadcrumbSwitcherLoading = {
  loading: true,
  resources: [],
};

export const breadcrumbSwitcherLoaded = {
  loading: false,
  resources: resourceList,
};

export const breadcrumbSwitcherLoadedWithPagination = {
  ...breadcrumbSwitcherLoaded,
  currentPage: 2,
  totalPages: 3,
};

export const breadcrumbBar = {
  data: {
    resource,
    breadcrumbItems: breadcrumbItems.items,
    isSwitchable: false,
  },
};

export const breadcrumbBarSwithcable = {
  data: {
    resource,
    breadcrumbItems: breadcrumbItems.items,
    isSwitchable: true,
  },
};
