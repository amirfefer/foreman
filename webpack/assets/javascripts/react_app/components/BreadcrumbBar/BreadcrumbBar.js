import React from 'react';
import PropTypes from 'prop-types';
import Breadcrumb from './components/Breadcrumb';
import BreadcrumbSwitcher from './components/BreadcrumbSwitcher';

class BreadcrumbBar extends React.Component {
  render() {
    const {
      data: { breadcrumbItems, isSwitchable, resource },
      resourceSwitcherItems,
      isLoadingResources,
      isSwitcherOpen,
      toggleSwitcher,
      closeSwitcher,
      loadSwitcherResourcesByResource,
      currentPage,
      totalPages,
    } = this.props;

    const isTitle = breadcrumbItems.length === 1;
    return (
      <div className="breadcrumb-bar">
        <Breadcrumb title items={breadcrumbItems} isTitle={isTitle}>
          {isSwitchable && (
            <BreadcrumbSwitcher
              open={isSwitcherOpen}
              isLoadingResources={isLoadingResources}
              resources={resourceSwitcherItems}
              onTogglerClick={() => toggleSwitcher()}
              onOverlayHide={() => closeSwitcher()}
              onOverlayEnter={() => loadSwitcherResourcesByResource(resource)}
              onNextPageClick={() =>
                 loadSwitcherResourcesByResource(resource, { page: Number(currentPage) + 1 }) }
              onPrevPageClick={() =>
                 loadSwitcherResourcesByResource(resource, { page: Number(currentPage) - 1 }) }
              currentPage={currentPage}
              totalPages={totalPages}
            />
          )}
        </Breadcrumb>
        {!isTitle && <hr className="breadcrumb-line" />}
      </div>
    );
  }
}

BreadcrumbBar.propTypes = {
  data: PropTypes.shape({
    isSwitchable: PropTypes.bool,
    resource: PropTypes.shape({
      url: PropTypes.string,
      controller: PropTypes.string,
      action: PropTypes.string,
      nameField: PropTypes.string,
    }),
    breadcrumbItems: PropTypes.arrayOf(PropTypes.shape({
      caption: PropTypes.string.isRequired,
      url: PropTypes.string.isRequired,
    })),
  }),
  resourceSwitcherItems: BreadcrumbSwitcher.propTypes.resources,
  isLoadingResources: PropTypes.bool,
  isSwitcherOpen: PropTypes.bool,
  toggleSwitcher: PropTypes.func,
  closeSwitcher: PropTypes.func,
  loadSwitcherResourcesByResource: PropTypes.func,
};

BreadcrumbBar.defaultProps = {
  data: {
    breadcrumbItems: [],
    isSwitchable: false,
  },
  resourceSwitcherItems: [],
  isLoadingResources: false,
  isSwitcherOpen: false,
  toggleSwitcher: () => null,
  closeSwitcher: () => null,
  loadSwitcherResourcesByResource: () => null,
};

export default BreadcrumbBar;
