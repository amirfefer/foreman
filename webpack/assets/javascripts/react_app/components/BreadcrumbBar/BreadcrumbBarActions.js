import API from '../../API';

import {
  BREADCRUMB_BAR_TOGGLE_SWITCHER,
  BREADCRUMB_BAR_CLOSE_SWITCHER,
  BREADCRUMB_BAR_RESOURCES_REQUEST,
  BREADCRUMB_BAR_RESOURCES_SUCCESS,
  BREADCRUMB_BAR_RESOURCES_FAILURE,
} from './BreadcrumbBarConstants';

export const toggleSwitcher = () => ({
  type: BREADCRUMB_BAR_TOGGLE_SWITCHER,
});

export const closeSwitcher = () => ({
  type: BREADCRUMB_BAR_CLOSE_SWITCHER,
});

export const loadSwitcherResourcesByResource = (resource, options = {}) => (dispatch) => {
  const {
    reosurceUrl, nameField, switcherItemUrl,
  } = resource;
  const { page = 1 } = options;

  const beforeRequest = () =>
    dispatch({
      type: BREADCRUMB_BAR_RESOURCES_REQUEST,
      payload: { resource, options },
    });

  const onRequestSuccess = response =>
    dispatch({ type: BREADCRUMB_BAR_RESOURCES_SUCCESS, payload: formatResults(response) });

  const onRequestFail = error =>
    dispatch({ type: BREADCRUMB_BAR_RESOURCES_FAILURE, payload: error });

  const formatResults = ({ data }) => {
    const switcherItems = Object.values(data.results).map((x, i) =>
      (data.results instanceof Array ?
        { name: x[nameField], url: switcherItemUrl.replace(':id', x.id) } :
        { name: x[i][nameField], url: switcherItemUrl.replace(':id', x[i].id) }));
    return {
      items: switcherItems,
      page: data.page,
      pages: Number(data.total) / Number(data.per_page),
    };
  };
  beforeRequest();

  return API.get(reosurceUrl, {}, { page }).then(onRequestSuccess, onRequestFail);
};
