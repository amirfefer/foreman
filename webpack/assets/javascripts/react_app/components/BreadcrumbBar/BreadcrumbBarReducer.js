import Immutable from 'seamless-immutable';

import {
  BREADCRUMB_BAR_TOGGLE_SWITCHER,
  BREADCRUMB_BAR_RESOURCES_REQUEST,
  BREADCRUMB_BAR_RESOURCES_SUCCESS,
  BREADCRUMB_BAR_RESOURCES_FAILURE,
} from './BreadcrumbBarConstants';

const initialState = Immutable({
  resourceSwitcherItems: [],
  isLoading: false,
  isSwitcherOpen: false,
  requestError: null,
});

export default (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case BREADCRUMB_BAR_RESOURCES_REQUEST:
      return state
        .set('resourceSwitcherItems', [])
        .set('requestError', null)
        .set('isLoading', true);

    case BREADCRUMB_BAR_RESOURCES_SUCCESS:
      return state
        .set('resourceSwitcherItems', payload)
        .set('requestError', null)
        .set('isLoading', false);

    case BREADCRUMB_BAR_RESOURCES_FAILURE:
      return state
        .set('resourceSwitcherItems', [])
        .set('requestError', payload)
        .set('isLoading', false);

    case BREADCRUMB_BAR_TOGGLE_SWITCHER:
      return state.set('isSwitcherOpen', !state.isSwitcherOpen);

    default:
      return state;
  }
};
