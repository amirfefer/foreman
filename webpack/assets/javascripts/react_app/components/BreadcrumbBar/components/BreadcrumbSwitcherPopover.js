import React from 'react';
import PropTypes from 'prop-types';
import { Popover, ListGroup, ListGroupItem, Pager } from 'patternfly-react';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';
import './BreadcrumbSwitcherPopover.scss';

const BreadcrumbSwitcherPopover = ({
  resources,
  onNextPageClick,
  onPrevPageClick,
  loading,
  currentPage,
  totalPages,
  ...props
}) => (
  <Popover className="breadcrumb-switcher-popover" {...props}>
    {loading ? (
      <div>Loading...</div>
    ) : (
      <ListGroup className="scrollable-list">
        {resources.map((item) => {
          const { onClick, url, name } = item;
          return (
            <ListGroupItem
              className="no-border"
              key={`id-${name}`}
              href={url}
              onClick={onClick}
            >
              <EllipsisWithTooltip>{name}</EllipsisWithTooltip>
            </ListGroupItem>
          );
        })}
      </ListGroup>
    )}
    <Pager
      className="pager-sm"
      messages={{ nextPage: __('Next'), previousPage: __('Previous') }}
      onNextPage={onNextPageClick}
      onPreviousPage={onPrevPageClick}
      disablePrevious={currentPage === 1}
      disableNext={currentPage === Math.ceil(totalPages)}
    />
  </Popover>
);

BreadcrumbSwitcherPopover.propTypes = {
  ...Popover.propTypes,
  loading: PropTypes.bool,
  resources: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    href: PropTypes.string,
    onClick: PropTypes.func,
  })),
};

BreadcrumbSwitcherPopover.defaultProps = {
  loading: false,
  resources: [],
};

export default BreadcrumbSwitcherPopover;
