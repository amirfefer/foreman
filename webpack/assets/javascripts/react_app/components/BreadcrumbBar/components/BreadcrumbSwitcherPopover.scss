@import '../../../common/colors';

.popover-content {
    max-height: $switcher-max-height;
    overflow-y: auto;
    max-width: $switcher-max-width;
    min-width: $switcher-max-width;

  .no-border {
    border: none;
  }
}

#breadcrumb-switcher-popover {
    z-index: 1000 !important;
}
