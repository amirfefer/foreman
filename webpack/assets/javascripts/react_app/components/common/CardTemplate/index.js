import React from 'react';
import CardItem from './CardItem';

import { addGlobalFill } from '../../common/Fill/GlobalFill';

const CardTemplate = ({ content, header, slotID, cardID }) => {
  addGlobalFill(
    slotID,
    cardID,
    <CardItem content={content} header={header} />,
    200
  );
};

export default CardTemplate;
