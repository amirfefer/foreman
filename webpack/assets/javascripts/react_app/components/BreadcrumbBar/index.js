import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import * as actions from './BreadcrumbBarActions';
import reducer from './BreadcrumbBarReducer';

import BreadcrumbBar from './BreadcrumbBar';

// map state to props
const mapStateToProps = ({ breadcrumbBar }) => ({
  resourceSwitcherItems: breadcrumbBar.resourceSwitcherItems,
  isSwitcherOpen: breadcrumbBar.isSwitcherOpen,
  isLoadingResources: breadcrumbBar.isLoadingResources,
  requestError: breadcrumbBar.requestError,
});

// map action dispatchers to props
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

// export reducers
export const reducers = { breadcrumbBar: reducer };

// export connected component
export default connect(mapStateToProps, mapDispatchToProps)(BreadcrumbBar);
