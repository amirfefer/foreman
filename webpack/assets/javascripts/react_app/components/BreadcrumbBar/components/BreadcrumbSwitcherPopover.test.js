import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

import BreadcrumbSwitcherPopover from './BreadcrumbSwitcherPopover';
import { breadcrumbSwitcherLoading, breadcrumbSwitcherLoaded } from '../BreadcrumbBar.fixtures';

const fixtures = {
  'render loading state': { id: 'some-id', ...breadcrumbSwitcherLoading },
  'render resources list': { id: 'some-id', ...breadcrumbSwitcherLoaded },
};

describe('BreadcrumbSwitcherPopover', () => testComponentSnapshotsWithFixtures(BreadcrumbSwitcherPopover, fixtures));
