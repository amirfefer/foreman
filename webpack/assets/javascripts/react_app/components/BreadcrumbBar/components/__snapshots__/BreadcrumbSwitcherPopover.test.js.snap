// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbSwitcherPopover render loading state 1`] = `
<Popover
  bsClass="popover"
  className="breadcrumb-switcher-popover"
  id="some-id"
  placement="right"
>
  <div>
    Loading...
  </div>
</Popover>
`;

exports[`BreadcrumbSwitcherPopover render resources list 1`] = `
<Popover
  bsClass="popover"
  className="breadcrumb-switcher-popover"
  id="some-id"
  placement="right"
>
  <ListGroup
    bsClass="list-group"
    className="scrollable-list"
  >
    <ListGroupItem
      bsClass="list-group-item"
      className="no-border"
      href="#"
      key="id-Host 1"
      listItem={false}
    >
      <EllipisWithTooltip>
        Host 1
      </EllipisWithTooltip>
    </ListGroupItem>
    <ListGroupItem
      bsClass="list-group-item"
      className="no-border"
      href="#"
      key="id-Host 2"
      listItem={false}
    >
      <EllipisWithTooltip>
        Host 2
      </EllipisWithTooltip>
    </ListGroupItem>
    <ListGroupItem
      bsClass="list-group-item"
      className="no-border"
      href="#"
      key="id-Host 3 with a very long name"
      listItem={false}
    >
      <EllipisWithTooltip>
        Host 3 with a very long name
      </EllipisWithTooltip>
    </ListGroupItem>
    <ListGroupItem
      bsClass="list-group-item"
      className="no-border"
      key="id-Host 4"
      listItem={false}
      onClick={[Function]}
    >
      <EllipisWithTooltip>
        Host 4
      </EllipisWithTooltip>
    </ListGroupItem>
    <ListGroupItem
      bsClass="list-group-item"
      className="no-border"
      href="#"
      key="id-Host 5"
      listItem={false}
    >
      <EllipisWithTooltip>
        Host 5
      </EllipisWithTooltip>
    </ListGroupItem>
  </ListGroup>
</Popover>
`;
